simulation/chart.py

```diff
@@ -1,232 +1,398 @@
+        y_label: Optional[str] = None,
+        y_range: Optional[Tuple[float, float]] = None,
+        y_log: bool = False,
+    ) -> ChartWidget:
+        """Add a chart with legend for plots as a row
+
+        Args:
+            title (str, optional): Title to display on the top. Defaults to `None`.
+            height (int): Prefered height scale factor.Defaults to `200`.
+            legend_width (int): Legend width. Defaults to `300`.
+            x_label (str, optional): Label text to display under x axis. Defaults to `None`.
+            x_range (Tuple[float,float], optional): Constant range(min,max value) of x axis. For dynamic range, set to `None`. Defaults to `None`.
+            x_log (bool): Use logarithmic scale to display x axis. Defaults to `False`.
+            y_label (str, optional): Label text to display next to y axis. Defaults to `None`.
+            y_range (Tuple[float,float], optional): Constant range(min,max value) of y axis. For dynamic range, set to `None`. Defaults to `None`.
+            y_log (bool): Use logarithmic scale for y axis. Defaults to `False`.
+
+        Returns:
+            ChartWidget: a chart to define and draw plots
+        """
+
+        chart = self._layout.addPlot(title=title)
+        chart.setPreferredHeight(height)
+        legend = chart.addLegend(offset=None)
+        vb: pg.ViewBox = self._layout.addViewBox()
+        vb.setFixedWidth(legend_width)
+        legend.setParentItem(vb)
+        legend.anchor((0, 0), (0, 0))
+        chart.setClipToView(True)
+        chart.setLogMode(x=x_log, y=y_log)
 
-def close_video_process(video_process: Popen):
-    """Waits until encoding sub process is finished data in the stdin
+        if x_label is not None:
+            chart.setLabel("bottom", x_label)
+        if x_range is not None:
+            chart.setXRange(*x_range)
 
-    Args:
-        video_process (Process): Vide (ffmpeg) sub process
-    """
-    video_process.stdin.close()  # pyright: ignore [reportOptionalMemberAccess]
-    video_process.wait()
+        if y_label is not None:
+            chart.setLabel("left", y_label)
+        if y_range is not None:
+            chart.setYRange(*y_range)
+
+        self._layout.nextRow()
+
+        return ChartWidget(chart)
+
+    def render(self):
+        """Render chart changes to UI or animation file."""
+        self._app.processEvents()  # type: ignore
+
+        if self._save:
+            # Renders a frame by capturing plots drawn on the layout.Re-scales if required based on the size.
+            qimage = self._layout.grab().toImage()
+
+            qimage = qimage.convertToFormat(
+                QtGui.QImage.Format_RGB888, QtCore.Qt.AutoColor  # type: ignore
+            )
+
+            # May have to rescale (HiDPI displays, etc)
+            if (qimage.width(), qimage.height()) != self._init_size:
+                qimage = (
+                    QtGui.QPixmap.fromImage(qimage)  # type: ignore
+                    .scaled(
```

Comment:
This is considered a bit of an anti-pattern in OOP though. I personally don't mind stuff like this as I think it's clear what's going on. 

@tkornuta-semiotic @aasseman, do either of you have a strong opinion about this block stylistically?

Reply:
The Qt library is extremely un-pythonnic anyway, so I don't have very high expectations for any good pattern to emerge here.
I suppose another way of doing it would be to assign each method result to `qimage`? That would make it a bit more pythonic.

Reply:
Yup. That would work. Would you want to change to that style, or is it not worth the effort in your opinion?

Reply:
Yeah, I don't have any problem here - just a sequence of image processing operations resulting in rescaling.

Not sure why we need to change the formats in this `if` (and not for the other sizes), but I guess @tumaysem did that for a reason.

Let's close this one and process with PR.